src/utils.ts

```diff
@@ -1254,3 +1255,28 @@ export async function request(
     req.end();
   });
 }
+
+/**
+ * A custom AbortController that aborts after a specified timeout.
+ *
+ * If `timeout` is undefined or \<=0, the abort controller never aborts.
+ *
+ * This class provides two benefits over the built-in AbortSignal.timeout() method.
+ * - This class provides a mechanism for cancelling the timeout
+ * - This class supports infinite timeouts by interpreting a timeout of 0 as infinite.  This is
+ *    consistent with existing timeout options in the Node driver (serverSelectionTimeoutMS, for example).
+ * @internal
+ */
+export class TimeoutController extends AbortController {
+  constructor(
+    timeout?: number,
+    private timeoutId = timeout && timeout > 0 ? setTimeout(() => this.abort(), timeout) : null
```

Comment:
abort accepts a reason, should we be giving it the error we want to reject with here?

Reply:
No - notice the abort handling still happens at the callsite.  I think that's a cleaner SOC - the timeout controller is responsible with timing out and the callsite is responsible for handling the timeout.  Also, some errors rely on state that occur at the time of the error.  Server selection timeouts, for example, include a topology description of the topology at the time of the timeout.